src/core/announcements/validation.ts

```diff
@@ -0,0 +1,254 @@
+  if (!isRecord(obj)) return false;
+  if (obj["AnnouncementType"] != AnnouncementType.Reaction) return false;
+  if (!isDSNPUserId(obj["fromId"])) return false;
+  if (!isValidEmoji(obj["emoji"])) return false;
+  if (!isDSNPAnnouncementId(obj["inReplyTo"])) return false;
+
+  return true;
+};
+
+/**
+ * isProfileAnnouncement() is a type check for ProfileAnnouncement
+ *
+ * @param obj - The object to check
+ * @returns True if the object is a ProfileAnnouncement, otherwise false
+ */
+const isProfileAnnouncement = (obj: unknown): obj is ProfileAnnouncement => {
+  if (!isRecord(obj)) return false;
+  if (obj["AnnouncementType"] != AnnouncementType.Profile) return false;
+  if (!isDSNPUserId(obj["fromId"])) return false;
+  if (!isString(obj["url"])) return false;
+  if (!isString(obj["contentHash"])) return false;
+
+  return true;
+};
+
+/**
+ * isAnnouncement() is a type check for Announcement
+ *
+ * @param obj - The object to check
+ * @returns True if the object is a Announcement, otherwise false
+ */
+const isAnnouncement = (obj: unknown): obj is Announcement => {
+  if (!isRecord(obj)) return false;
+  if (!isAnnouncementType(obj["AnnouncementType"])) return false;
+
+  const validators = {
+    [AnnouncementType.GraphChange]: isGraphChangeAnnouncement,
+    [AnnouncementType.Broadcast]: isBroadcastAnnouncement,
+    [AnnouncementType.Reply]: isReplyAnnouncement,
+    [AnnouncementType.Reaction]: isReactionAnnouncement,
+    [AnnouncementType.Profile]: isProfileAnnouncement,
+  };
+
+  return validators[obj["AnnouncementType"]](obj);
+};
+
+/**
+ * isSignedAnnouncement() is a type check for SignedAnnouncements
+ *
+ * @param obj - The object to check
+ * @returns True if the object is a SignedAnnouncement, otherwise false
+ */
+export const isSignedAnnouncement = (obj: unknown): obj is SignedAnnouncement => {
+  if (!isRecord(obj)) return false;
+  if (!isString(obj["signature"])) return false;
+  return isAnnouncement(obj);
+};
+
+const validateKeccak256FileHash = async (
+  activityContent: ActivityContentAudio | ActivityContentImage | ActivityContentVideo
+): Promise<boolean> => {
+  let keccakHash: string | null = null;
+  if (Array.isArray(activityContent["hash"])) {
+    for (const hash in activityContent["hash"]) {
+      if (!isActivityContentHash(hash)) return false;
+      if (hash["algorithm"] == "keccak256") keccakHash = hash["value"];
+    }
+  } else {
+    if (!isActivityContentHash(activityContent["hash"])) return false;
+    if (activityContent["hash"]["algorithm"] == "keccak256") keccakHash = activityContent["hash"]["value"];
+  }
+
+  if (keccakHash === null) return true;
+
+  let fileUrl: string | null = null;
+  if (typeof activityContent["url"] === "string") {
+    fileUrl = activityContent["url"];
+  } else {
+    fileUrl = activityContent["url"]["href"];
+  }
+
+  const fileContents = await fetch(fileUrl).then((res) => res.text());
+  return keccak256(fileContents) == keccakHash;
```

Comment:
I don't think we should be fetching these files. They could be very large, and if they are desired, they should be fetched and cached.

Instead we should expose the hash function (instead of requiring people import js-sha3) (We have a PT story for that: https://www.pivotaltracker.com/story/show/178904143)